# Машинное обучение, ФЭН ВШЭ

# Практическое задание 2
(Основано на материалах курса машинного обучения на фкн)

## Общая информация

Дата выдачи: 21.09.2017

Срок сдачи: 8.10.2017 23:59MSK

### О задании

Практическое задание 1 посвящено изучению основных библиотек для анализа данных, а также линейных моделей и методов их обучения. Вы научитесь:
 * применять библиотеку NumPy;
 * подготавливать данные для обучения линейных моделей;
 * обучать линейную, Lasso и Ridge-регрессии при помощи модуля scikit-learn;
 * реализовывать обычный и стохастический градиентные спуски;
 * обучать линейную регрессию для произвольного функционала качества.
 

### Оценивание и штрафы

Каждая из задач имеет определенную «стоимость» (указана в скобках около задачи). Максимально допустимая оценка за работу — 10 баллов. Кроме того, некоторые из заданий являются опциональными (необязательными), однако за их выполнение можно получить дополнительные баллы.


Задание выполняется самостоятельно. «Похожие» решения считаются плагиатом и все задействованные студенты (в том числе те, у кого списали) не могут получить за него больше 0 баллов (подробнее о плагиате см. на странице курса). Если вы нашли решение какого-то из заданий (или его часть) в открытом источнике, необходимо указать ссылку на этот источник в отдельном блоке в конце Вашей работы (скорее всего вы будете не единственным, кто это нашел, поэтому чтобы исключить подозрение в плагиате, необходима ссылка на источник). 

Неэффективная реализация кода может негативно отразиться на оценке.


### Формат сдачи
Для сдачи задания переименуйте получившийся файл \*.ipynb в соответствии со следующим форматом: *HW1_Username.ipynb*, где *Username* — Ваша фамилия и инициалы на латинице (например, *HW1_IvanovII.ipynb*). Далее отправьте этот файл на почту курса, указав соответствующую тему (см. вики-страничку курча).

## Библиотеки для анализа данных

### NumPy

Во всех заданиях данного раздела запрещено использовать циклы  и list comprehensions. Под вектором и матрицей в данных заданиях понимается одномерный и двумерный numpy.array соответственно.

In [1]:
import numpy as np

In [2]:
import math

**1. (0.3 балла)** Реализуйте функцию, возвращающую максимальный элемент в векторе x среди элементов, перед которыми стоит нулевой. Для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответом является 5. Если нулевых элементов нет, функция должна возвращать None.


In [ ]:
def max_element(arr):
    # Your code here

**2. (0.3 балла)** Реализуйте функцию, принимающую на вход матрицу и некоторое число и возвращающую ближайший к числу элемент матрицы. Например: для X = np.arange(0,10).reshape((2, 5)) и v = 3.6 ответом будет 4.

Ресурс: https://stackoverflow.com/questions/2566412/find-nearest-value-in-numpy-array

In [ ]:
def nearest_value(X, v):
    idx = (np.abs(X-v)).argmin()
    return array[idx]

**3. (0.3 балла)** Реализуйте функцию scale(X), которая принимает на вход матрицу и масштабирует каждый ее столбец (вычитает выборочное среднее и делит на стандартное отклонение). Убедитесь, что в функции не будет происходить деления на ноль. Протестируйте на случайной матрице (для её генерации можно использовать, например, функцию [numpy.random.randint](http://docs.scipy.org/doc/numpy/reference/generated/numpy.random.randint.html)).

Ресурс: https://stackoverflow.com/questions/1735025/how-to-normalize-a-numpy-array-to-within-a-certain-range

In [72]:
mu, sigma = 10, 1 # mean and standard deviation
s = np.random.normal(mu, sigma, 100)
s.resize((10,10))
def scale(X):
    X /= np.max(np.abs(X),axis=0)
scale(s)
s

array([[ 0.8268583 ,  0.92621354,  0.8662165 ,  0.84741073,  0.78536464,
         0.95679358,  0.94320551,  0.86010064,  0.83592358,  0.89291499],
       [ 0.81404281,  0.89539163,  1.        ,  1.        ,  0.76618056,
         0.89581251,  0.76017926,  0.87333299,  1.        ,  1.        ],
       [ 0.92451384,  0.85525256,  0.65810577,  0.74593229,  0.95203599,
         0.84979842,  0.99586438,  0.74287043,  0.7885883 ,  0.74781173],
       [ 0.92714506,  0.82484116,  0.96223608,  0.74208122,  0.6629004 ,
         0.93840401,  1.        ,  0.75947238,  0.92846523,  0.86740636],
       [ 0.96236711,  0.82881511,  0.82677142,  0.77721417,  0.83364602,
         0.78810828,  0.87597695,  1.        ,  0.83975073,  0.99093039],
       [ 0.93138576,  0.7949837 ,  0.80421552,  0.82794583,  0.67564296,
         0.85218041,  0.97245037,  0.7986723 ,  0.78638549,  0.86954227],
       [ 0.83469872,  0.8627669 ,  0.91476413,  0.81362024,  0.92890211,
         0.92764259,  0.92738936,  0.92027499

**4. (0.3 балла)** Реализуйте функцию, которая для заданной матрицы находит:
 - определитель
 - след
 - наименьший и наибольший элементы
 - норму Фробениуса
 - собственные числа
 - обратную матрицу

Для тестирования сгенерируйте матрицу с элементами из нормального распределения $\mathcal{N}$(10,1)

In [11]:
mu, sigma = 10, 1 # mean and standard deviation
s = np.random.normal(mu, sigma, 100)
s.resize((10,10))
s

array([[  8.8801814 ,   9.79942952,   8.49766413,  11.48068356,
          9.87508646,  10.37898803,   9.66848319,   9.98301456,
         10.21130671,  10.41351382],
       [  9.59354334,  10.85470916,   9.46354486,   7.96189439,
         11.40243826,   8.74713554,  10.3622494 ,   8.16408309,
          9.2272987 ,   9.91666943],
       [ 10.75444797,   8.35239442,   9.84233398,   9.72769734,
          8.82958526,   8.51601763,   9.94656525,   9.91027024,
         11.25779363,   9.49796318],
       [ 10.52772379,   9.31914558,   9.5293264 ,   9.70620953,
          7.93786159,  10.32462774,  10.58239148,   8.22195777,
         12.5715985 ,   9.47073932],
       [ 11.15646243,  11.18346224,  12.12205161,  10.06201349,
          8.99939229,   8.229124  ,   9.47047359,   8.88096805,
         10.52563294,   9.90782296],
       [ 10.37739615,  11.51603403,  10.21573116,   9.31458664,
         11.34396578,  11.13098611,  10.0520422 ,   9.97559652,
          9.3282144 ,  10.7678136 ],
       [ 1

In [37]:
def get_stats(X):
    print("определитель:", np.linalg.det(X))
    print("след:", np.trace(X, offset=0, axis1=0, axis2=1, dtype=None, out=None))
    print("наибольший элемент:", np.amax(X))
    print("наименьший элемент:", np.amin(X))
    print("норма Фробениуса:", np.linalg.norm(X, ord=None, axis=None, keepdims=False))
    print("собственные числа:", np.linalg.eigvals(X))
    print("обратная матрица:", np.linalg.inv(X))

In [38]:
get_stats(s)

определитель: 21978.4763919
след: 100.972609754
наибольший элемент: 12.571598499
наименьший элемент: 7.93786158581
норма Фробениуса: 100.952585808
собственные числа: [ 100.35186679+0.j           -2.68133753+0.j           -1.99101899+0.j
    0.26472567+2.26791071j    0.26472567-2.26791071j
   -0.46645717+0.6781742j    -0.46645717-0.6781742j
    1.80234738+1.51783304j    1.80234738-1.51783304j    2.09186772+0.j        ]
обратная матрица: [[ -6.73553949e-01  -8.90831245e-01  -1.38906171e-01   5.69921906e-01
    2.94535189e-01   2.43335251e+00  -1.26663683e+00  -3.51193150e-01
   -9.93594137e-01   9.29403515e-01]
 [  1.91883129e-01   1.31069771e-01  -3.51886987e-01  -8.44036748e-03
    2.07190732e-01  -3.83317055e-01   2.06860542e-01   1.77738984e-01
   -7.94737928e-03  -1.61686028e-01]
 [ -3.69652090e-02   1.16001364e-01   1.62492278e-01  -1.69248419e-01
   -4.63547209e-03  -3.65147732e-01   2.07089656e-01   2.38693942e-02
    3.38060153e-01  -2.39156654e-01]
 [ -1.81474708e-01  -7.775902

**5. (0.3 балла)** Повторите 100 раз следующий эксперимент: сгенерируйте две матрицы размера 10×10 из стандартного нормального распределения, перемножьте их (как матрицы) и найдите максимальный элемент. Какое среднее значение по экспериментам у максимальных элементов? 95-процентная квантиль?

In [69]:
mu, sigma = 0, 1 # mean and standard deviation
list0 = []
average = 0
for exp_num in range(100): 
    m1 = np.random.normal(mu, sigma, 100)
    m1.resize((10,10))
    m2 = np.random.normal(mu, sigma, 100)
    m2.resize((10,10))
    result = np.dot(m1, m2)
    average += np.amax(result) / 100
    list0.append(np.amax(result))
print("Среднее значение:", average)
print("95-процентная квантиль:", np.percentile(list0, 0.95))

Среднее значение: 8.28983056365
95-процентная квантиль: 5.13150185505


## Линейная регрессия

В этой части мы разберемся с линейной регрессией, способами её обучения и измерением качества ее прогнозов. 

Будем рассматривать данные по авиарейсам в США за январь-апрель 2008 года для предсказания времени задержки отправления рейса в минутах (DepDelay). Отметим, что под задержкой подразумевается не только опоздание рейса относительно планируемого времени вылета, но и отправление до планируемого времени.


[Данные](https://www.dropbox.com/s/dvfitn93obn0rql/2008.csv?dl=0) и их [описание](http://stat-computing.org/dataexpo/2009/the-data.html)

### Подготовка данных

**6. (0.75 балла)** Считайте выборку из файла при помощи функции pd.read_csv и ответьте на следующие вопросы:
   - Имеются ли в данных пропущенные значения?
   - Сколько всего пропущенных элементов в таблице "объект-признак"?
   - Сколько объектов имеют хотя бы один пропуск?
   - Сколько признаков имеют хотя бы одно пропущенное значение?

In [3]:
import pandas as pd

In [4]:
df = pd.read_csv("2008.csv")
df.head()

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0  2008      6          18          3   2111.0        2055   2308.0   
1  2008      6           4          3   1426.0        1410   1730.0   
2  2008      8           3          7   1143.0        1145   1501.0   
3  2008      1          23          3   1141.0        1144   1323.0   
4  2008      5           4          7    815.0         820   1243.0   

   CRSArrTime UniqueCarrier  FlightNum        ...         TaxiIn  TaxiOut  \
0        2300            WN        242        ...            3.0      8.0   
1        1728            XE       2380        ...            4.0     12.0   
2        1520            WN       1769        ...            4.0      9.0   
3        1335            OO       3802        ...            4.0     19.0   
4        1300            WN        399        ...            4.0      8.0   

   Cancelled  CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0          0               NaN         0           NaN          NaN      NaN   
1          0               NaN         0           NaN          NaN      NaN   
2          0               NaN         0           NaN          NaN      NaN   
3          0               NaN         0           NaN          NaN      NaN   
4          0               NaN         0           NaN          NaN      NaN   

   SecurityDelay  LateAircraftDelay  
0            NaN                NaN  
1            NaN                NaN  
2            NaN                NaN  
3            NaN                NaN  
4            NaN                NaN  

[5 rows x 29 columns]

In [5]:
df.describe()

C:\Program Files\Anaconda3\lib\site-packages\numpy\lib\function_base.py:4116: RuntimeWarning: Invalid value encountered in percentile
  interpolation=interpolation)


Year         Month    DayofMonth     DayOfWeek       DepTime  \
count  70000.0  70000.000000  70000.000000  70000.000000  68601.000000   
mean    2008.0      6.380300     15.742300      3.925900   1331.025058   
std        0.0      3.419011      8.813738      1.995955    477.005973   
min     2008.0      1.000000      1.000000      1.000000      1.000000   
25%     2008.0      3.000000      8.000000      2.000000           NaN   
50%     2008.0      6.000000     16.000000      4.000000           NaN   
75%     2008.0      9.000000     23.000000      6.000000           NaN   
max     2008.0     12.000000     31.000000      7.000000   2400.000000   

         CRSDepTime       ArrTime    CRSArrTime     FlightNum  \
count  70000.000000  68444.000000  70000.000000  70000.000000   
mean    1324.217071   1478.600842   1492.904771   2214.752114   
std      463.231562    504.601176    482.071631   1963.606525   
min        1.000000      1.000000      1.000000      1.000000   
25%      926.000000           NaN   1115.000000    614.000000   
50%     1316.000000           NaN   1515.000000   1562.000000   
75%     1715.000000           NaN   1905.000000   3500.000000   
max     2359.000000   2400.000000   2400.000000   7828.000000   

       ActualElapsedTime        ...              Distance        TaxiIn  \
count       68418.000000        ...          70000.000000  68444.000000   
mean          126.960931        ...            724.508257      6.847189   
std            70.382109        ...            565.485009      4.884862   
min            19.000000        ...             31.000000      0.000000   
25%                  NaN        ...            324.000000           NaN   
50%                  NaN        ...            575.000000           NaN   
75%                  NaN        ...            950.000000           NaN   
max           686.000000        ...           4962.000000    165.000000   

            TaxiOut     Cancelled      Diverted  CarrierDelay  WeatherDelay  \
count  68594.000000  70000.000000  70000.000000  15253.000000  15253.000000   
mean      16.435621      0.020157      0.002443     15.862388      2.829935   
std       11.297027      0.140539      0.049365     38.596304     17.687256   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%             NaN      0.000000      0.000000           NaN           NaN   
50%             NaN      0.000000      0.000000           NaN           NaN   
75%             NaN      0.000000      0.000000           NaN           NaN   
max      317.000000      1.000000      1.000000   1146.000000    516.000000   

           NASDelay  SecurityDelay  LateAircraftDelay  
count  15253.000000   15253.000000       15253.000000  
mean      16.751918       0.078935          20.698944  
std       32.017588       2.063460          39.068022  
min        0.000000       0.000000           0.000000  
25%             NaN            NaN                NaN  
50%             NaN            NaN                NaN  
75%             NaN            NaN                NaN  
max      914.000000     175.000000         784.000000  

[8 rows x 24 columns]

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70000 entries, 0 to 69999
Data columns (total 29 columns):
Year                 70000 non-null int64
Month                70000 non-null int64
DayofMonth           70000 non-null int64
DayOfWeek            70000 non-null int64
DepTime              68601 non-null float64
CRSDepTime           70000 non-null int64
ArrTime              68444 non-null float64
CRSArrTime           70000 non-null int64
UniqueCarrier        70000 non-null object
FlightNum            70000 non-null int64
TailNum              69180 non-null object
ActualElapsedTime    68418 non-null float64
CRSElapsedTime       69991 non-null float64
AirTime              68418 non-null float64
ArrDelay             68418 non-null float64
DepDelay             68601 non-null float64
Origin               70000 non-null object
Dest                 70000 non-null object
Distance             70000 non-null int64
TaxiIn               68444 non-null float64
TaxiOut              68594 non-

In [7]:
a = len(df.index)-df.count()
a.head()
a.cumsum(axis=0, skipna=True)

Year                      0
Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                1399
CRSDepTime             1399
ArrTime                2955
CRSArrTime             2955
UniqueCarrier          2955
FlightNum              2955
TailNum                3775
ActualElapsedTime      5357
CRSElapsedTime         5366
AirTime                6948
ArrDelay               8530
DepDelay               9929
Origin                 9929
Dest                   9929
Distance               9929
TaxiIn                11485
TaxiOut               12891
Cancelled             12891
CancellationCode      81480
Diverted              81480
CarrierDelay         136227
WeatherDelay         190974
NASDelay             245721
SecurityDelay        300468
LateAircraftDelay    355215
dtype: int64

Ресурс: https://stackoverflow.com/questions/22257527/how-do-i-get-a-summary-count-of-missing-nan-data-by-column-in-pandas  
Имеются ли в данных пропущенные значения? Да  
Сколько всего пропущенных элементов в таблице "объект-признак"? 355215

In [9]:
a = pd.isnull(df)
b = [0] * 70000
for i in range(len(a.index)):
    for j in range(len(a.columns)):
        if a.iloc[i, j] == True and b[i] == 0:
            b[i] += 1
c = 0
for i in range(len(b)):
    c += b[i]
c

70000

Все 70000 объектов имеют хотя бы одно пропущенное значение.

In [45]:
a = pd.isnull(df)
b = [0] * len(a.columns)
for i in range(len(a.index)):
    for j in range(len(a.columns)):
        if (a.iloc[i, j] == True or math.isnan(df.iloc[i, j] == True)) and b[j] == 0:
            b[j] += 1
c = 0
for i in range(len(b)):
    c += b[i]
c

16

In [ ]:
16 признаков имеют хотя бы одно пропущенное значение

Как вы понимаете, также не имеет смысла рассматривать при решении поставленной задачи объекты с пропущенным значением целевой переменной. В связи с этим ответьте на следующие вопросы и выполните соответствующие действия:
- Имеются ли пропущенные значения в целевой переменной?
- Проанализируйте объекты с пропущенными значениями целевой переменной. Чем вызвано это явление? Что их объединяет? Можно ли в связи с этим, на ваш взгляд, исключить какие-то признаки из рассмотрения? Обоснуйте свою точку зрения.

Исключите из выборки объекты **с пропущенным значением целевой переменной и со значением целевой переменной, равным 0**, а также при необходимости исключите признаки в соответствии с вашим ответом на последний вопрос из списка и выделите целевую переменную в отдельный вектор, исключив её из матрицы "объект-признак".

In [11]:
d = 0
for i in range(len(a.index)):
    if math.isnan(df.iloc[i]["DepDelay"]) == True:
        d += 1 
print("Пропущено", d, "значений в целевой переменной")

Пропущено 1399 значений в целевой переменной


In [13]:
df = df[df.DepDelay != 0]

In [14]:
df = df.dropna(axis=0, how='any', thresh=None, subset=["DepDelay"], inplace=False)

In [32]:
list = []
for i in range(len(df.index)):
    cell = df.iloc[i]["DepDelay"]
    if math.isnan(cell) == True:
        list.append(df.index[i])

In [33]:
list

[11,
 28,
 31,
 34,
 70,
 71,
 75,
 78,
 85,
 92,
 94,
 100,
 103,
 104,
 112,
 120,
 125,
 128,
 133,
 135,
 137,
 144,
 171,
 179,
 194,
 200,
 201,
 204,
 234,
 236,
 239,
 244,
 257,
 267,
 269,
 275,
 282,
 288,
 290,
 302,
 308,
 310,
 318,
 332,
 352,
 375,
 398,
 415,
 423,
 428,
 429,
 430,
 438,
 439,
 445,
 446,
 469,
 480,
 490,
 500,
 526,
 538,
 539,
 542,
 553,
 554,
 556,
 564,
 572,
 582,
 583,
 584,
 590,
 591,
 604,
 635,
 648,
 654,
 675,
 686,
 690,
 696,
 697,
 712,
 718,
 720,
 721,
 737,
 802,
 811,
 816,
 817,
 821,
 823,
 830,
 838,
 839,
 848,
 865,
 876,
 882,
 898,
 943,
 965,
 975,
 992,
 1019,
 1039,
 1048,
 1066,
 1068,
 1072,
 1089,
 1094,
 1100,
 1111,
 1112,
 1127,
 1135,
 1139,
 1150,
 1156,
 1169,
 1186,
 1189,
 1193,
 1198,
 1204,
 1227,
 1261,
 1263,
 1264,
 1267,
 1268,
 1272,
 1280,
 1307,
 1308,
 1315,
 1318,
 1335,
 1345,
 1349,
 1352,
 1379,
 1386,
 1396,
 1401,
 1406,
 1418,
 1424,
 1426,
 1431,
 1438,
 1445,
 1446,
 1448,
 1456,
 1472,
 147

In [35]:
df = df.drop(list)

**7. (1 балл)** Обратите внимание, что признаки DepTime, CRSDepTime, ArrTime, CRSArrTime приведены в формате hhmm, в связи с чем будет не вполне корректно рассматривать их как вещественные.

Преобразуйте каждый признак FeatureName из указанных в пару новых признаков FeatureName\_Hour, FeatureName\_Minute, разделив каждое из значений на часы и минуты. Не забудьте при этом исключить исходный признак из выборки. В случае, если значение признака отсутствует, значения двух новых признаков, его заменяющих, также должны отсутствовать. 

Например, признак DepTime необходимо заменить на пару признаков DepTime_Hour, DepTime_Minute. При этом, например, значение 155 исходного признака будет преобразовано в значения 1 и 55 признаков DepTime_Hour, DepTime_Minute соответственно.

In [36]:
df["DepTime_Hour"] = np.nan
df["DepTime_Minute"] = np.nan
df["CRSDepTime_Hour"] = np.nan
df["CRSDepTime_Minute"] = np.nan
df["ArrTime_Hour"] = np.nan
df["ArrTime_Minute"] = np.nan
df["CRSArrTime_Hour"] = np.nan
df["CRSArrTime_Minute"] = np.nan

In [37]:
import math
def string_length(str1):
    count = 0
    for char in str1:
        count += 1
    return count
def convert(X):
    for i in range(len(df.index)):
        cell = df.iloc[i][X]
        if math.isnan(cell) == False:
            if cell == 0:
                cell = ''.join(('0000', cell))
            cell = int(cell)
            cell = str(cell)
            if string_length(cell) == 1:
                cell = ''.join(('000', cell))
            if string_length(cell) == 2:
                cell = ''.join(('00', cell))
            if string_length(cell) == 3:
                cell = ''.join(('0', cell))
            df.set_value(i, X + '_Hour', float(cell[:2]))
            df.set_value(i, X + '_Minute', float(cell[2:]))

In [38]:
convert("DepTime")
convert("CRSDepTime")
convert("ArrTime")
convert("CRSArrTime")

**8. (0.5 балла)** Некоторые из признаков, отличных от целевой переменной, могут оказывать чересчур значимое влияние на прогноз, поскольку по своему смыслу содержат большую долю информации о значении целевой переменной. Изучите описание датасета и исключите признаки, сильно коррелирующие с ответами. Ваш выбор признаков для исключения из выборки обоснуйте. Кроме того, исключите признаки TailNum и Year.

In [56]:
#Diverted, CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay
t = df.corr(method='pearson', min_periods=1)
list2 = []
for i in range(len(t.columns)):
    if t.iloc[13][i] > 0.5 or t.iloc[13][i] < -0.5:
        list2.append(t.columns[i])
list2

['ArrDelay', 'DepDelay', 'CarrierDelay', 'LateAircraftDelay']

Выше указан список признаков, которые по моему мнению нужно удалить (кроме DepDelay): они в достаточной степени скоррелированы с целевой переменной.

In [61]:
df = df.drop(['ArrDelay', 'CarrierDelay', 'LateAircraftDelay'], axis=1)
df = df.drop(["TailNum", "Year"], axis=1)

Приведем данные к виду, пригодному для обучения линейных моделей. Для этого вещественные признаки надо отмасштабировать, а категориальные — привести к числовому виду. Также надо устранить пропуски в данных.

**9*. (1 бонусный балл) ** В первую очередь поймем, зачем необходимо применять масштабирование. Следующие ячейки с кодом построят гистограммы для 3 вещественных признаков выборки.

In [ ]:
X['DepTime_Hour'].hist(bins=20)

In [ ]:
X['TaxiIn'].hist(bins=20)

In [ ]:
X['FlightNum'].hist(bins=20)

Какую проблему вы наблюдаете на этих графиках? Как масштабирование поможет её исправить?

**10. (1 балл)** Вдобавок к масштабированию, некоторые из признаков в нашем датасете являются категориальными. Типичным подходом к работе с ними является бинарное, или [one-hot-кодирование](https://en.wikipedia.org/wiki/One-hot).


Реализуйте функцию transform_data, которая принимает на вход DataFrame с признаками и выполняет следующие шаги:
1. Замена пропущенных значений на нули для вещественных признаков и на строки 'nan' для категориальных.
2. Масштабирование вещественных признаков с помощью [StandardScaler](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html).
3. One-hot-кодирование категориальных признаков с помощью [DictVectorizer](http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.DictVectorizer.html) или функции [pd.get_dummies](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html).

Метод должен возвращать преобразованный DataFrame, который должна состоять из масштабированных вещественных признаков и закодированных категориальных (исходные признаки должны быть исключены из выборки).

In [ ]:
def transform_data(data):
    # Your code here

Примените функцию transform_data к данным. Сколько признаков получилось после преобразования?

In [ ]:
# Your code here

**11. (0.75 балла)** Разбейте выборку и вектор целевой переменной на обучение и контроль в отношении 70/30 (для этого можно использовать, например, функцию [train_test_split](http://scikit-learn.org/stable/modules/generated/sklearn.cross_validation.train_test_split.html)). 

In [ ]:
# Your code here

### Scikit-learn

<img src = "https://pp.vk.me/c4534/u35727827/93547647/x_d31c4463.jpg">
Теперь, когда мы привели данные к пригодному виду, попробуем решить задачу при помощи метода наименьших квадратов. Напомним, что данный метод заключается в оптимизации функционала $MSE$:

$$MSE(X, y) = \frac{1}{l} \sum_{i=1}^l (<w, x_i> - y_i)^2 \to \min_{w},$$

где $\{ (x_i, y_i ) \}_{i=1}^l$ — обучающая выборка, состоящая из $l$ пар объект-ответ.

Заметим, что решение данной задачи уже реализовано в модуле sklearn в виде класса [LinearRegression](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html#sklearn.linear_model.LinearRegression).

**12. (0.5 балла)** Обучите линейную регрессию на 1000 объектах из обучающей выборки и выведите значения $MSE$ и $R^2$ на этой подвыборке и контрольной выборке (итого 4 различных числа). Проинтерпретируйте полученный результат — насколько качественные прогнозы строит полученная модель? Какие проблемы наблюдаются в модели?

**Подсказка**: изучите значения полученных коэффициентов $w$, сохраненных в атрибуте coef_ объекта LinearRegression.

In [ ]:
# Your code here

Для решения описанных вами в предыдущем пункте проблем используем L1- или L2-регуляризацию, тем самым получив Lasso и Ridge регрессии соответственно и изменив оптимизационную задачу одним из следующих образов:
$$MSE_{L1}(X, y) = \frac{1}{l} \sum_{i=1}^l (<w, x_i> - y_i)^2 + \alpha ||w||_1 \to \min_{w},$$
$$MSE_{L2}(X, y) = \frac{1}{l} \sum_{i=1}^l (<w, x_i> - y_i)^2 + \alpha ||w||_2^2 \to \min_{w},$$

где $\alpha$ — коэффициент регуляризации. Один из способов его подбора заключается в переборе некоторого количества значений и оценке качества на кросс-валидации для каждого из них, после чего выбирается значение, для которого было получено наилучшее качество.

**13. (1 балл)** Обучите линейные регрессии с L1- и L2-регуляризатором, подобрав лучшее значение параметра регуляризации из списка alpha_grid при помощи кросс-валидации c 5 фолдами на тех же 1000 объектах, что и в п.12. Выведите значения $MSE$ и $R^2$ на обучающей и контрольной выборках. Удалось ли решить указанные вами ранее проблемы?

Для выполнения данного задания вам могут понадобиться реализованные в библиотеке объекты [LassoCV](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LassoCV.html), [RidgeCV](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.RidgeCV.html) и [KFold](http://scikit-learn.org/stable/modules/generated/sklearn.cross_validation.KFold.html).


In [ ]:
# Your code here

### Градиентный спуск

В предыдущем разделе мы использовали существующие реализации методов обучения линейной регрессии с регуляризацией и без. Тем не менее, подобные реализации, как правило, имеются лишь для ограниченного набора стандартных методов. В частности, при выходе функционала качества за пределы стандартного множества необходимо самостоятельно реализовывать составляющие процесса решения оптимизационной задачи. Именно этому и посвящен данный раздел задания.

Пусть необходимо минимизировать следующий функционал (Mean Square Percentage Error — модифицированный [RMSPE](https://www.kaggle.com/c/rossmann-store-sales/details/evaluation)):
$$MSPE(\{x_i, y_i\}_{i=1}^l, \, w) = \frac{1}{l}\sum_{i=1}^l \left( \frac{y_i - \langle w, x_i \rangle }{y_i} \right)^2,$$

где $\{x_i, y_i\}_{i=1}^l$ — обучающая выборка, $w$ — вектор весов линейной модели. Будем также рассматривать функционал $MSPE$ с L2-регуляризацией:

$$MSPE(\{x_i, y_i\}_{i=1}^l, \, w) = \frac{1}{l}\sum_{i=1}^l \left( \frac{y_i - \langle w, x_i \rangle }{y_i} \right)^2 + ||w||_2^2.$$

**14. (0 баллов)** Добавьте к объектам обеих выборок из п. 11 единичный признак.

In [ ]:
# Your code here

**15. (1 балл)** Реализуйте функции, которые вычисляют:
 * прогнозы линейной модели;
 * функционал $MSPE$ и его градиент;
 * регуляризованный $MSPE$ и его градиент.

In [ ]:
# возвращает вектор прогнозов линейной модели с вектором весов w для выборки X
def make_pred(X, w):
    pass

In [ ]:
# возвращает значение функционала MSPE для выборки (X, y) и вектора весов w
def get_func(w, X, y):
    pass

In [ ]:
# возвращает градиент функционала MSPE для выборки (X, y) и вектора весов w
def get_grad(w, X, y):
    pass

In [ ]:
# возвращает значение регуляризованного функционала MSPE для выборки (X, y) и вектора весов w
def get_reg_func(w, X, y):
    pass

In [ ]:
# возвращает градиент регуляризованного функционала MSPE для выборки (X, y) и вектора весов w
def get_reg_grad(w, X, y):
    pass

**16. (1 балл)** Реализуйте метод градиентного спуска для описанных функционалов ($MSPE$ и его регуляризованный вариант). Функция должна принимать следующие параметры:
 - X — матрица "объект-признак";
 - y — вектор целевой переменной;
 - w0 — начальное значение вектора весов;
 - step_size — значение темпа обучения;
 - max_iter — максимальное число итераций;
 - eps — значение, используемое в критерии останова;
 - is_reg — бинарный параметр, принимает значение True в случае наличия регуляризации функционала, False — в противном случае.
 
Процесс должен быть остановлен, если выполнено хотя бы одно из следующих условий:
 - было выполнено заданное количество итераций max_iter;
 - евклидова норма разности векторов $w$ на соседних итерациях стала меньше, чем eps.

Функция должна возвращать полученный в результате оптимизации вектор $w$ и список значений функционала на каждой итерации.

In [ ]:
def grad_descent(X, y, step_size, max_iter, eps, is_reg):
    # Your code here

Обучите линейную регрессию с функционалом $MSPE$ на обучающей выборке при помощи метода градиентного спуска и изобразите кривые зависимости значения функционала от номера итерации для различных:
 * значений размера шага из набора [0.001, 1, 10];
 * способов начальной инициализации вектора весов (нули, случайные веса).

Проанализируйте полученные результаты — влияют ли данные параметры на скорость сходимости и итоговое качество? Если да, то как?

In [ ]:
# Your code here

**17. (0.5 балла)** Обучите линейную регрессию с функционалом MSPE и его регуляризованным вариантом на обучающей выборке при помощи метода градиентного спуска и изобразите кривые зависимости значения функционала от номера итерации. Исследуйте зависимость скорости сходимости от наличия регуляризации. Обоснуйте, почему так происходит.

In [ ]:
# Your code here

Метод градиентного спуска может быть весьма трудозатратен в случае большого размера обучающей выборки. Поэтому часто используют метод стохастического градиентного спуска, где на каждой итерации выбирается случайный объект из обучающей выборки и обновление весов происходит только по этому объекту. 

**18*. (1 бонусный балл)**  Реализуйте метод стохастического градиентного спуска (SGD) для описанных функционалов ($MSPE$ и его регуляризованный вариант). Функция должна иметь параметры и возвращаемое значение, аналогичные оным функции grad\_descent из п.21. Кроме того, должен использоваться аналогичный критерий останова.

In [ ]:
def sgd(X, y, step_size, max_iter, eps, is_reg):
    # Your code here

Обучите линейную регрессию с функционалом $MSPE$ и его регуляризованным вариантом на обучающей выборке при помощи метода стохастического градиентного спуска, подобрав при этом размер шага, при котором метод будет сходиться. Нарисуйте график сходимости. Выведите значения $MSPE, MSE, R^2$ на контрольной выборке.

In [ ]:
# Your code here

**19*. (0.5 бонусного балла)** Аналогично п.17 исследуйте зависимость скорости сходимости метода SGD от наличия регуляризации. Обоснуйте, почему так происходит.

In [ ]:
# Your code here

**20. (0.5 балла)** Обучите стандартную линейную регрессию с функционалом качества MSE на обучающей выборке и выведите значение MSPE полученного решения на контрольной выборке. Как оно соотносится с аналогичным результатом для решения, полученного в п.17? Почему?

In [ ]:
# Your code here